src/PushNotifications.php

```diff
@@ -94,38 +144,101 @@ public function publish($interests, $publishRequest) {
-      $response = $e->GetResponse();
-      $parsedResponse = json_decode($response->GetBody());
-      $badJSON = $parsedResponse == null;
-      if (
-        $badJSON ||
-        !ARRAY_KEY_EXISTS('error', $parsedResponse) ||
-        !ARRAY_KEY_EXISTS('description', $parsedResponse)
-      ) {
-        throw new \Exception("An unexpected server error has occurred");
+    $path = '/publish_api/v1/instances/$instanceId/publishes/interests';
+    $pathParams = [
+      '$instanceId' => $this->options["instanceId"]
+    ];
+    $response = $this->makeRequest("POST", $path, $pathParams, $publishRequest);
+
+    if ($response == null) {
+      throw new \Exception("An unexpected server error has occurred");
+    }
+
+    return $response;
+  }
+
+  public function publishToUsers($userIds, $publishRequest) {
+    if (!is_array($userIds)) {
+      throw new \Exception("'userIds' must be an array");
+    }
+    if (count($userIds) == 0) {
+      throw new \Exception("Publishes must target at least one user");
+    }
+    if (count($userIds) > PushNotifications::MAX_USERS) {
+      throw new \Exception("Number of user ids exceeds maximum of " . PushNotifications::MAX_USERS);
+    }
+    if(!is_array($publishRequest)) {
+      throw new \Exception("'publishBody' must be an array");
+    }
+
+    foreach($userIds as $userId) {
+      if (!is_string($userId)) {
+        throw new \Exception("User id \"$userId\" is not a string");
+      }
+      if (strlen($userId) > PushNotifications::MAX_USER_ID_LENGTH) {
+        throw new \Exception("User id \"$userId\" is longer than the maximum length of " . PushNotifications::MAX_USER_ID_LENGTH . " chars.");
+      }
+      if (strlen($userId) == 0) {
+        throw new \Exception("User ids cannot be the empty string");
       }
-      throw new \Exception("{$parsedResponse->error}: {$parsedResponse->description}");
     }
 
-    $parsedResponse = json_decode($response->GetBody());
-    if ($parsedResponse == null) {
+    $publishRequest['users'] = $userIds;
+    $path = '/publish_api/v1/instances/$instanceId/publishes/users';
+    $pathParams = [
+      '$instanceId' => $this->options["instanceId"]
+    ];
+    $response = $this->makeRequest("POST", $path, $pathParams, $publishRequest);
+
+    if ($response == null) {
       throw new \Exception("An unexpected server error has occurred");
     }
 
-    return $parsedResponse;
+    return $response;
+  }
+
+  public function deleteUser($userId) {
+    if (!is_string($userId)) {
+      throw new \Exception("User id must be a string");
+    }
+    if (strlen($userId) == 0) {
+      throw new \Exception("User id cannot be the empty string");
+    }
+    if (strlen($userId) > PushNotifications::MAX_USER_ID_LENGTH) {
+      throw new \Exception("User id \"$userId\" is longer than the maximum length of " . PushNotifications::MAX_USER_ID_LENGTH . " chars.");
+    }
+
+    $path = '/user_api/v1/instances/$instanceId/users/$userId';
```

Comment:
This looks a bit too clever for me, I think.
It sort of feels like it would be `"/user_api/v1/instances/$instanceId/users/$userId"` (with quotes). But that would be wrong as it would need to be
`"/user_api/v1/instances/$instanceId/users/$urlencodedUserId"`.

Maybe if the paths were defined as consts outside the function, it would be more obvious/forcing that you'd need to use the `pathParams` field. I don't know, I'm just back from holidays, so feel free to ignore this :o